src/py21cmfast/src/SpinTemperatureBox.c

```diff
@@ -1,2760 +1,1932 @@
 
-        // let's initialize an array of redshifts (z'') corresponding to the
-        // far edge of the dz'' filtering shells
-        // and the corresponding minimum halo scale, sigma_Tmin,
-        // as well as an array of the frequency integrals
-        LOG_SUPER_DEBUG("beginning loop over R_ct");
+    //converts the grid emissivity unit to per cm-3
+    if(flag_options_ts->USE_HALO_FIELD){
+        consts->volunit_inv = pow(CMperMPC,-3); //changes to emissivity per cm-3
+    }
+    else{
+        consts->volunit_inv = cosmo_params_ts->OMb * RHOcrit * pow(CMperMPC,-3);
+    }
 
-        for (R_ct=0; R_ct<global_params.NUM_FILTER_STEPS_FOR_Ts; R_ct++){
-            if (R_ct==0){
-                prev_zpp = zp;
-                prev_R = 0;
-            }
-            else{
-                prev_zpp = zpp_edge[R_ct-1];
-                prev_R = R_values[R_ct-1];
-            }
+    consts->growth_zp = dicke(zp);
+    consts->hubble_zp = hubble(zp);
+    consts->dgrowth_dzp = ddicke_dz(zp);
+    consts->dt_dzp = dtdz(zp);
 
-            zpp_edge[R_ct] = prev_zpp - (R_values[R_ct] - prev_R)*CMperMPC / drdz(prev_zpp); // cell size
-            zpp = (zpp_edge[R_ct]+prev_zpp)*0.5; // average redshift value of shell: z'' + 0.5 * dz''
+    LOG_DEBUG("Set zp consts xr %.2e Tr %.2e Ts %.2e xa %.2e xc %.2e cm %.2e",consts->xray_prefactor,consts->Trad,
+                consts->Ts_prefactor,consts->xa_tilde_prefactor,consts->xc_inverse,consts->dcomp_dzp_prefactor);
+    LOG_DEBUG("Nb %.2e la %.2e vi %.2e D %.2e H %.2e dD %.2e dt %.2e",consts->Nb_zp,consts->lya_star_prefactor,
+                consts->volunit_inv,consts->growth_zp,consts->hubble_zp,consts->dgrowth_dzp,consts->dt_dzp);
+}
 
-            zpp_for_evolve_list[R_ct] = zpp;
-            if (R_ct==0){
-                dzpp_for_evolve = zp - zpp_edge[0];
-            }
-            else{
-                dzpp_for_evolve = zpp_edge[R_ct-1] - zpp_edge[R_ct];
-            }
-            zpp_growth[R_ct] = dicke(zpp);
-            if (flag_options->USE_MINI_HALOS){
-                Mcrit_atom_interp_table[R_ct] = atomic_cooling_threshold(zpp);
-            }
+//All the cell-dependent stuff needed to calculate Ts
+struct Box_rad_terms{
+    double dxion_dt;
+    double dxheat_dt;
+    double dxlya_dt;
+    double dstarlya_dt;
+    double dstarLW_dt;
+    double dstarlya_cont_dt;
+    double dstarlya_inj_dt;
+    double delta;
+    double prev_Ts;
+    double prev_Tk;
+    double prev_xe;
+};
+
+//outputs from the Ts calculation, to go into new boxes
+struct Ts_cell{
+    double Ts;
+    double x_e;
+    double Tk;
+    double J_21_LW;
+};
+
+//Function for calculating the Ts box outputs quickly by using pre-calculated constants
+//  as much as possible
+//TODO: unpack the structures at the start?
+//  BUT surely any compiler should optimise it....
+//TODO: Move more constants out of the loop
+struct Ts_cell get_Ts_fast(float zp, float dzp, struct Ts_zp_consts *consts, struct Box_rad_terms *rad){
```

Comment:
It's just a function which returns a struct (not pointer). Ages ago brad optimised the `get_Ts` function but it was all just in the main `ComputeTs` which I've put into `get_Ts_fast`. Here I found it convenient to define a struct representing all the quantities we needed from a single cell, and have it return that rather than managing pointers which go in/out of the function